### PR TITLE
[FLINK-14037]Use same userCodeClassLoader as jobManager in jobClient to fix input/…

### PR DESCRIPTION
**Pull request by liupc**
## What is the purpose of the change

When user include `flink-shaded-hadoop-2` classes into user jar, and use a custom inputformat class (e.g. `TeraInputFormat`) which extending the hadoop inputformat(e.g. `FileInputFormat`), then in current version, there will be a deserilization error. 

That's because the classloader in jobManager is childFirst(with a whitelist that always loaded by parent loader), however, the classloader in client is parentFirst.

This PR is trying to keep the above two classloaders to be consistent, thus to avoid such odd serde exceptions.

```
Caused by: org.apache.flink.runtime.client.JobSubmissionException: Failed to submit JobGraph.
	at org.apache.flink.client.program.rest.RestClusterClient.lambda$submitJob$8(RestClusterClient.java:382)
	at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:870)
	at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:852)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at org.apache.flink.runtime.concurrent.FutureUtils.lambda$retryOperationWithDelay$8(FutureUtils.java:263)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:561)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:929)
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.flink.runtime.rest.util.RestClientException: [Internal server error., <Exception on server side:
org.apache.flink.runtime.client.JobSubmissionException: Failed to submit job.
	at org.apache.flink.runtime.dispatcher.Dispatcher.lambda$internalSubmitJob$2(Dispatcher.java:333)
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:797)
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:44)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.RuntimeException: org.apache.flink.runtime.client.JobExecutionException: Could not set up JobManager
	at org.apache.flink.util.function.CheckedSupplier.lambda$unchecked$0(CheckedSupplier.java:36)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
	... 6 more
Caused by: org.apache.flink.runtime.client.JobExecutionException: Could not set up JobManager
	at org.apache.flink.runtime.jobmaster.JobManagerRunner.<init>(JobManagerRunner.java:152)
	at org.apache.flink.runtime.dispatcher.DefaultJobManagerRunnerFactory.createJobManagerRunner(DefaultJobManagerRunnerFactory.java:83)
	at org.apache.flink.runtime.dispatcher.Dispatcher.lambda$createJobManagerRunner$5(Dispatcher.java:375)
	at org.apache.flink.util.function.CheckedSupplier.lambda$unchecked$0(CheckedSupplier.java:34)
	... 7 more
Caused by: org.apache.flink.runtime.client.JobExecutionException: Cannot initialize task 'DataSource (at org.apache.flink.api.scala.ExecutionEnvironment.createInput(ExecutionEnvironment.scala:390) (org.apache.flink.api.scala.hadoop.mapreduce.HadoopInpu)': Loading the input/output formats failed: org.apache.flink.api.scala.hadoop.mapreduce.HadoopInputFormat@2e179f3e
	at org.apache.flink.runtime.executiongraph.ExecutionGraphBuilder.buildGraph(ExecutionGraphBuilder.java:218)
	at org.apache.flink.runtime.executiongraph.ExecutionGraphBuilder.buildGraph(ExecutionGraphBuilder.java:106)
	at org.apache.flink.runtime.scheduler.LegacyScheduler.createExecutionGraph(LegacyScheduler.java:207)
	at org.apache.flink.runtime.scheduler.LegacyScheduler.createAndRestoreExecutionGraph(LegacyScheduler.java:184)
	at org.apache.flink.runtime.scheduler.LegacyScheduler.<init>(LegacyScheduler.java:176)
	at org.apache.flink.runtime.scheduler.LegacySchedulerFactory.createInstance(LegacySchedulerFactory.java:70)
	at org.apache.flink.runtime.jobmaster.JobMaster.createScheduler(JobMaster.java:275)
	at org.apache.flink.runtime.jobmaster.JobMaster.<init>(JobMaster.java:265)
	at org.apache.flink.runtime.jobmaster.factories.DefaultJobMasterServiceFactory.createJobMasterService(DefaultJobMasterServiceFactory.java:98)
	at org.apache.flink.runtime.jobmaster.factories.DefaultJobMasterServiceFactory.createJobMasterService(DefaultJobMasterServiceFactory.java:40)
	at org.apache.flink.runtime.jobmaster.JobManagerRunner.<init>(JobManagerRunner.java:146)
	... 10 more
Caused by: java.lang.Exception: Loading the input/output formats failed: org.apache.flink.api.scala.hadoop.mapreduce.HadoopInputFormat@2e179f3e
	at org.apache.flink.runtime.jobgraph.InputOutputFormatVertex.initInputOutputformatContainer(InputOutputFormatVertex.java:156)
	at org.apache.flink.runtime.jobgraph.InputOutputFormatVertex.initializeOnMaster(InputOutputFormatVertex.java:60)
	at org.apache.flink.runtime.executiongraph.ExecutionGraphBuilder.buildGraph(ExecutionGraphBuilder.java:214)
	... 20 more
Caused by: java.lang.RuntimeException: Deserializing the input/output formats failed: unread block data
	at org.apache.flink.runtime.jobgraph.InputOutputFormatContainer.<init>(InputOutputFormatContainer.java:68)
	at org.apache.flink.runtime.jobgraph.InputOutputFormatVertex.initInputOutputformatContainer(InputOutputFormatVertex.java:153)
	... 22 more
Caused by: java.lang.IllegalStateException: unread block data
	at java.io.ObjectInputStream$BlockDataInputStream.setBlockDataMode(ObjectInputStream.java:2783)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1605)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2287)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2211)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2069)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
	at org.apache.flink.util.InstantiationUtil.deserializeObject(InstantiationUtil.java:576)
	at org.apache.flink.util.InstantiationUtil.deserializeObject(InstantiationUtil.java:562)
	at org.apache.flink.util.InstantiationUtil.deserializeObject(InstantiationUtil.java:550)
	at org.apache.flink.util.InstantiationUtil.readObjectFromConfig(InstantiationUtil.java:511)
	at org.apache.flink.runtime.operators.util.TaskConfig.getStubWrapper(TaskConfig.java:288)
	at org.apache.flink.runtime.jobgraph.InputOutputFormatContainer.<init>(InputOutputFormatContainer.java:66)
	... 23 more

End of exception on server side>]
	at org.apache.flink.runtime.rest.RestClient.parseResponse(RestClient.java:389)
	at org.apache.flink.runtime.rest.RestClient.lambda$submitRequest$3(RestClient.java:373)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:952)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926)
```

## Brief change log

  - *Modify client userCodeClassLoader to keep the classloader consistent both in client and jobManager*

## Verifying this change

This change is already covered by existing tests, such as *ClassLoaderITCase/PackageProgramTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? ( no)

### CODE REVIEW

File: flink-clients/src/main/java/org/apache/flink/client/program/JobWithJars.java [hunk 1]
```diff
@@ -131,14 +133,22 @@ public static void checkJarFile(URL jar) throws IOException {
 		}
 	}
 
-	public static ClassLoader buildUserCodeClassLoader(List<URL> jars, List<URL> classpaths, ClassLoader parent) {
+	public static ClassLoader buildUserCodeClassLoader(
+		List<URL> jars, List<URL> classpaths, ClassLoader parent, Configuration conf) {
```
Comment (zhuzhurk) [inline flink-clients/src/main/java/org/apache/flink/client/program/JobWithJars.java]: Each param should be in a new line.

File: flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgram.java [hunk 2]
```diff
@@ -202,7 +228,9 @@ public PackagedProgram(File jarFile, List<URL> classpaths, @Nullable String entr
 		// now that we have an entry point, we can extract the nested jar files (if any)
 		this.extractedTempLibraries = jarFileUrl == null ? Collections.emptyList() : extractContainedLibraries(jarFileUrl);
 		this.classpaths = classpaths;
-		this.userCodeClassLoader = JobWithJars.buildUserCodeClassLoader(getAllLibraries(), classpaths, getClass().getClassLoader());
+
+		this.userCodeClassLoader = JobWithJars.buildUserCodeClassLoader(
+			getAllLibraries(), classpaths, getClass().getClassLoader(), conf);
```
Comment (zhuzhurk) [inline flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgram.java]: Each param should be in a new line.

Comment (liupc) [reply flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgram.java]: Done

File: flink-clients/src/main/java/org/apache/flink/client/program/JobWithJars.java (hunk 1 above)
Comment (liupc) [reply flink-clients/src/main/java/org/apache/flink/client/program/JobWithJars.java]: Done

**Comment by aljoscha** [comment]
@TisonKun could you maybe have a look at this PR. I think this will clash with https://issues.apache.org/jira/browse/FLINK-14113.

Regarding this PR itself: I think adding `Configuration` as a parameter is not the right approach because now we have a lot of new code places where we just create an empty `Configuration`, that doesn't seem right. As an alternative, the client could create a classloader and pass it to the packaged program or the other code paths that need it.

**Comment by TisonKun** [comment]
Thanks for reminding me @aljoscha. For the conflict part I notice that it is only related with `JobWithJars.buildUserCodeClassLoader` which is a static method so that FLINK-14113 can live with this commit. Adaption is needed within change to the location of `buildUserCodeClassLoader`.

**Comment by TisonKun** [comment]
For the PR itself, I second @aljoscha that it doesn't seem right we pass a lot of empty `Configuration` to fit in a new function. @liupc Could you try to achieve the function by creating a classloader in client side and pass it to the packaged program or the other code paths that need it?

**Comment by liupc** [comment]
@TisonKun @aljoscha Thanks for reply, I will update as commented.

**Comment by liupc** [comment]
@aljoscha @TisonKun I find it's hard to make the classloader outside the packagedProgram, there are lots of logic and deps in PackagedProgram to build the classpaths for the classsloader. So I just get the `classLoaderResoloverOrder` and `alwaysParentFirstLoaderPatterns` at client and pass them to the constructor arguments of the PackagedProgram.

**Comment by aljoscha** [comment]
I think now we're in a similar situation. We have parameters for `PackagedProgram` where we often pass in `null`.  This hides the fact that sometimes we use the defaults, without using configuration and sometimes we don't. If we force all users of `PackageProgram` to provide a `ClassLoader` it's the problem of that code to provide the right class loader. This is better separation of concerns.

**Comment by liupc** [comment]
@aljoscha Thanks, I've update the code.

**Comment by liupc** [comment]
Any more suggestions? @TisonKun @aljoscha

**Comment by TisonKun** [comment]
could you sync with FLINK-13749 a.k.a  https://github.com/apache/flink/pull/10113?

**Comment by aljoscha** [comment]
Now that #10113 is merged, do we still need parts of this PR?

Also, I'm sorry for the double work here between the two PRs. 😓

**Comment by liupc** [comment]
@aljoscha  I also updated the `classloader.parent-first-patterns.default` confs in my PR, I encoutered failure without these confs in my testing, I will try to check them next week and put another PR for this if it's a real problem.
I will close this PR for now.

**Comment by aljoscha** [comment]
thanks! please ping me then

**Pull request closed without merging**